fix: use tokio spawn / interval.tick(), make nats metric names clearer, fix tests sharing environment variables (temp_env)

keivenchang opened the pull request:
#### Overview:

Follow-up to https://github.com/ai-dynamo/dynamo/pull/2480 with improvements and feedback

drt runtime::spawn, interval.tick(), refactored NATS metrics organization, and fixed test parallelization issues

#### Details:

- Use drt runtime::spawn instead of thread::spawn, and interval.tick() instead of sleep(...). Note that the blocking problem was lready fixed in PR 2480 where I removed std::thread::spawn, used tokio::runtime::Handle::try_current()). Graham and Olson said to use tokio spawn/interval.tick, which I'm doing here. 
- Split NATS metrics into separate `nats_client` and `nats_service` modules for clearer separation of concerns
- Enhanced NATS service metrics with `service_name` and component-specific labels for better observability
- Fixed test parallelization failures by ensuring `uptime_seconds` metric is consistently registered
- Improved duplicate metric registration handling with intelligent error categorization
- Updated test helpers to use `temp_env` for proper environment variable isolation

#### Where should the reviewer start?

- `lib/runtime/src/component.rs` - Check the improved error handling for duplicate metrics
- `lib/runtime/src/metrics/prometheus_names.rs` - See the new module structure for NATS metrics
- `lib/runtime/src/distributed.rs` - Review the uptime_seconds metric registration logic

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

Fixes DIS-445

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Expanded NATS client metrics (bytes, messages, connects, connection state).
  - Introduced NATS service metrics (avg processing time, totals, active services/endpoints) with consistent labels (including service name).
  - Uptime metrics now tracked even when system HTTP endpoints are disabled.

- Refactor
  - Split metrics into two namespaces: nats_client_* and nats_service_* (old nats_* renamed).
  - Streamlined background metrics scraping with fixed-interval runtime tasks and a shorter polling interval.

- Bug Fixes
  - More robust handling of duplicate Prometheus metric registrations to prevent startup conflicts.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
refactor: split NATS metrics into client and service modules, improve test isolation

- Refactored prometheus_names to separate nats_client and nats_service modules
- Added service_name and component labels to NATS service metrics for better observability
- Fixed test parallelization issues by ensuring uptime_seconds metric is always registered
- Updated SystemStatusState to handle duplicate metric registrations gracefully
- Modified create_test_drt helpers to use temp_env for proper environment isolation
- Updated all tests to use new nats_client/nats_service metric names
- Improved error logging to distinguish between expected duplicate metrics and real errors
- Cleaned up component metric scraping method naming and documentation

keivenchang commented on lib/runtime/src/metrics/prometheus_names.rs:
```diff
@@ -32,69 +32,73 @@ pub mod labels {
+pub mod nats_client {
     /// Prefix for all NATS client metrics
-    pub const PREFIX: &str = "nats_";
+    pub const PREFIX: &str = "nats_client_";
```
Actually I got it working with a macro. Something like:
```
    macro_rules! nats_service_prefix {
        ($name:expr) => {
            concat!("nats_service_", $name)
        };
    }
```
and then use that throughout the mod.

Sadly, it's much more involved that C++ where I can do the following.
```
const std::string a = "a";
const std::string b = a + a; // This is valid, but the concatenation happens at runtime
```

keivenchang replied:
Refactored in the next PR.

rmccorm4 commented on lib/runtime/src/component.rs:
```diff
@@ -575,12 +580,27 @@ impl Namespace {
-                component.service_name(),
-                err
-            );
+        if let Err(err) = component.start_scraping_nats_service_component_metrics() {
```
> Doesn't happen in tests

Not needed for this PR specifically - but can we add a test case that would make it happen, if not there already? Should hopefully reduce some heartburn from encountering it only in the wild but not in tests

keivenchang replied:
Will address duplicate metrics -> re-use metrics, in 0.5.0.

keivenchang committed:
refactor: improve NATS metrics and test infrastructure

- Add macro-based prefix generation for nats_client and nats_service metrics
- Update component scraping interval from 777ms to 873ms
- Convert failing integration tests to async tokio::test format
- Improve error handling in distributed runtime start time initialization
- Fix test helper functions to use proper async runtime context

keivenchang commented on lib/runtime/src/distributed.rs:
```diff
@@ -146,7 +147,18 @@ impl DistributedRuntime {
             }
         } else {
-            tracing::debug!("Health and system status server is disabled via DYN_SYSTEM_ENABLED");
+            // System server HTTP is disabled, but still create the state for metrics
```
I'll fix this design flaw in 0.5.0.

keivenchang commented on lib/runtime/src/metrics.rs:
```diff
@@ -584,7 +584,7 @@ mod test_helpers {
     pub fn create_test_drt() -> crate::DistributedRuntime {
```
This is deleted, everything will use create_test_drt_async now.

keivenchang commented on lib/runtime/src/metrics.rs:
```diff
@@ -939,9 +939,9 @@ mod test_metricsregistry_prefixes {
-    fn test_hierarchical_prefixes_and_parent_hierarchies() {
-        let drt = super::test_helpers::create_test_drt();
+    #[tokio::test]
+    async fn test_hierarchical_prefixes_and_parent_hierarchies() {
```
Changed all the tests to use tokio::test, for consistency and ease of debugging (1 type instead of 2).

keivenchang committed:
refactor: remove unused create_test_drt function from metrics test helpers

keivenchang committed:
refactor: rename NATS metric macros from _prefix to _name for clarity

rmccorm4 commented on lib/runtime/src/component.rs:
```diff
@@ -267,18 +271,19 @@ impl Component {
         const NATS_TIMEOUT_AND_INITIAL_DELAY_MS: std::time::Duration =
             std::time::Duration::from_millis(300);
-        const MAX_DELAY_MS: std::time::Duration = std::time::Duration::from_millis(4700);
+        const MAX_DELAY_MS: std::time::Duration = std::time::Duration::from_millis(873);
```
Just for my understanding - what would happen if this was set to `1000` (1 second)? I think this random number raises more questions than necessary when scanning the code, maybe something like `500` or `1000` would be easier for people to scan past without stopping to wonder why it's set to such a value (even with the comment describing it). Not sure if this is just me though.

keivenchang replied:
I have a test that blasts a bunch of messages (client->server), waits a second, and then checks the NATS stats. That's why :) But, I'll make sure to make a comment about this.

rmccorm4 reviewed (approved):
Approving to unblock - looking forward to the other follow-ups from comments on test case. dummy gauge, max delay number, etc. - thanks for the improvements!
